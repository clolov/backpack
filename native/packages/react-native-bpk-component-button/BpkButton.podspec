require 'json'

package = JSON.parse(File.read(File.join(__dir__, 'package.json')))

Pod::Spec.new do |s|
  s.name             = package['name']
  s.version          = package['version']
  s.summary          = package['description']
  s.license          = package['license']
  s.homepage         = 'https://github.com/skyscanner/backpack'
  s.authors          = package['author']
  s.source           = { :git => 'https://github.com/skyscanner/backpack.git', :tag => "v#{s.version}" }
  s.source_files     = 'src/*[^\-test].js'
  s.dependency         'React'
end
