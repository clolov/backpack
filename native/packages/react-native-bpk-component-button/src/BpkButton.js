/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2017 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

 import {
   View,
   Text,
   TouchableHighlight,
 } from 'react-native';
 import React from 'react';
 import PropTypes from 'prop-types';
 import LinearGradient from 'react-native-linear-gradient';

 import styles from './BpkButton-styles';

 const getTextStyle = ({ type, large, disabled }) => {
   const textStyle = [styles.base.text];
   if (styles[type].text) {
     textStyle.push(styles[type].text);
   }
   if (large) {
     textStyle.push(styles.large.text);
   }
   if (disabled) {
     textStyle.push(styles.disabled.text);
   }
   return textStyle;
 };

 const getButtonStyle = ({ type, disabled, selected }) => {
   const btnStyle = [styles.base.button];
   if (styles[type].button) {
     btnStyle.push(styles[type].button);
   }
   if (selected) {
     btnStyle.push(styles.selected.button);
   }
   if (disabled) {
     btnStyle.push(styles.disabled.button);
   }
   return btnStyle;
 };

 const getGradientColors = ({ type, disabled, selected }) => {
   let stylesToApply = styles[type].gradientColors;
   if (selected) {
     stylesToApply = styles.selected.gradientColors;
   }
   if (disabled) {
     stylesToApply = styles.disabled.gradientColors;
   }
   return stylesToApply;
 };

 const BpkButton = (props) => {
   const {
     type,
     title,
     onPress,
     large,
     disabled,
     selected,
     children,
     ...rest
   } = props;
   // Note that TouchableHighlight isn't on Android, so TouchableFeedback
   // will need to be used to support it.
   return (
     <LinearGradient style={styles.base.container} colors={getGradientColors(props)}>
       <TouchableHighlight
         style={getButtonStyle(props)}
         disabled={disabled}
         selected={selected}
         onPress={onPress}
         underlayColor={styles.base.underlayColor}
         {...rest}
       >
         <View>
           <Text style={getTextStyle(props)}>{title}</Text>
           {children}
         </View>
       </TouchableHighlight>
     </LinearGradient>
   );
 };

 BpkButton.propTypes = {
   title: PropTypes.string,
   onPress: PropTypes.func.isRequired,
   type: PropTypes.string.isRequired,
   large: PropTypes.bool,
   disabled: PropTypes.bool,
   selected: PropTypes.bool,
   children: PropTypes.node,
 };

 BpkButton.defaultProps = {
   title: null,
   type: 'primary',
   large: false,
   disabled: false,
   selected: false,
   children: null,
 };

 export default BpkButton;
