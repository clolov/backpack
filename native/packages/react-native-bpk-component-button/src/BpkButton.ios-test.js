/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2017 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

 import { Image } from 'react-native';
 import React from 'react';
 import renderer from 'react-test-renderer';

 import BpkButton, { BUTTON_TYPES } from './BpkButton';

 jest.mock('react-native', () => {
   const reactNative = require.requireActual('react-native');
   jest
     .spyOn(reactNative.Platform, 'select')
     .mockImplementation(obj => obj.ios || obj.default);
   reactNative.Platform.OS = 'ios';

   return reactNative;
 });

 jest.mock('Image', () => 'Image');

 describe('iOS', () => {
   describe('BpkButton', () => {
     it('should render correctly', () => {
       const tree = renderer.create(
         <BpkButton title="Lorem ipsum" onPress={() => {}} />,
       ).toJSON();
       expect(tree).toMatchSnapshot();
     });

     it('should support the "large" property', () => {
       const tree = renderer.create(
         <BpkButton large title="Lorem ipsum" onPress={() => {}} />,
       ).toJSON();
       expect(tree).toMatchSnapshot();
     });

     it('should support the "selected" property', () => {
       const tree = renderer.create(
         <BpkButton selected title="Lorem ipsum" onPress={() => {}} />,
       ).toJSON();
       expect(tree).toMatchSnapshot();
     });

     it('should support the "disabled" property', () => {
       const tree = renderer.create(
         <BpkButton large title="Lorem ipsum" onPress={() => {}} />,
       ).toJSON();
       expect(tree).toMatchSnapshot();
     });

     it('should support having an icon as well as a title', () => {
       const tree = renderer.create(
         <BpkButton
           icon={<Image source="../rightarrow_360.png" />}
           title="Lorem ipsum"
           onPress={() => {}}
         />,
       ).toJSON();
       expect(tree).toMatchSnapshot();
     });

     it('should support having only an icon', () => {
       const tree = renderer.create(
         <BpkButton
           icon={<Image source="../rightarrow_360.png" />}
           onPress={() => {}}
         />,
       ).toJSON();
       expect(tree).toMatchSnapshot();
     });

     it('should support overwriting styles', () => {
       const tree = renderer.create(
         <BpkButton title="Lorem ipsum" onPress={() => {}} style={{ width: 100 }} />,
       ).toJSON();
       expect(tree).toMatchSnapshot();
     });

     BUTTON_TYPES.forEach((buttonType) => {
       it(`should render correctly with type="${buttonType}"`, () => {
         const tree = renderer.create(
           <BpkButton type={buttonType} title="Lorem ipsum" onPress={() => {}} />,
         ).toJSON();
         expect(tree).toMatchSnapshot();
       });
     });
   });
 });
