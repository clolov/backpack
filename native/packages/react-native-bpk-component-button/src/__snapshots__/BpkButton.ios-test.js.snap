// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`iOS BpkButton should render correctly 1`] = `
<View
  accessibilityComponentType="button"
  accessibilityLabel={undefined}
  accessibilityTraits={
    Array [
      "button",
    ]
  }
  accessible={true}
  collapsable={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Array [
        Object {},
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#007AFF",
            "fontSize": 18,
            "padding": 8,
            "textAlign": "center",
          },
        ]
      }
    >
      Lorem ipsum
    </Text>
  </View>
</View>
`;

exports[`iOS BpkButton should support overwriting styles 1`] = `
<View
  accessibilityComponentType="button"
  accessibilityLabel={undefined}
  accessibilityTraits={
    Array [
      "button",
    ]
  }
  accessible={true}
  collapsable={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Array [
        Object {},
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#007AFF",
            "fontSize": 18,
            "padding": 8,
            "textAlign": "center",
          },
        ]
      }
    >
      Lorem ipsum
    </Text>
  </View>
</View>
`;
